### PR TITLE
Cache multiple Android action button pushes if app is not running

### PR DESCRIPTION
**Pull request by mikepsinn**
## Description
Caches multiple Android action button pushes if the app is not currently running in the background. 

This was originally contributed by @nicoabie.  I just fixed some merge conflicts and added a forgotten 	import java.util.ArrayList.

## Related Issue
This fixes https://github.com/phonegap/phonegap-plugin-push/issues/1271

## Motivation and Context
Each android action button press overwrites the previously cached action if the app is not running in the background.

## How Has This Been Tested?
Manually using the steps outlined in https://github.com/phonegap/phonegap-plugin-push/issues/1271 and with `npm test`.

## Types of changes

- [X] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [X] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [X] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [X] All new and existing tests passed.

### CODE REVIEW

**Comment by mikepsinn** [comment]
@macdonst I'm not sure if you want to merge this given your previous comments, but it actually fixed my issue after I fixed some merge conflicts and added a missing import java.util.ArrayList.  So I thought I'd submit a PR.

**Comment by macdonst** [comment]
@mikepsinn this PR still has a concurrency issue when you are iterating through the list of messages. Also it seems weird that your notifications are now coming always coming in as an array. That would be a breaking change to the API which I don't really want to merge into a point release as it would mess everyone up and become incompatible with iOS.

**Comment by mikepsinn** [comment]
Thanks, @macdonst !  

By "coming in as an array" do you mean I'm sending them as an array?  They aren't coming as an array.  I'm adding them to an array here: https://github.com/phonegap/phonegap-plugin-push/pull/1272/files#diff-03cb1908228e8abe8c4784020250e805R252.

By "breaking change to the API" do you mean a change in the way we have to send the notifications from our server?  If so, why would that be the case?  I didn't have to change anything in the way I send the pushes from my server.  It just adds the button actions to an array on Android.  How would that affect iOS?

After writing those questions above, I realize you were probably talking about the changes to index.spec.js.  I don't really understand what was going on there with the arrays, but I've reverted those changes and the tests still pass with the changes to PushPlugin.java.  So if that's what you were referrring to, feel free to disregard my questions above.

Thanks! :D

**Comment by macdonst** [comment]
@mikepsinn yeah, I meant the changes to the tests. It looked like the messages came in as arrays of messages but that didn't jive with the code. 

The PR looks good except for the concurrency issue. I'll add comments directly to the code.

File: src/android/com/adobe/phonegap/push/PushPlugin.java [hunk 1]
```diff
@@ -31,7 +32,7 @@
 
     private static CallbackContext pushContext;
     private static CordovaWebView gWebView;
-    private static Bundle gCachedExtras = null;
+    private static ArrayList<Bundle> gCachedExtras = new ArrayList<Bundle>();
```
Comment (macdonst) [inline src/android/com/adobe/phonegap/push/PushPlugin.java]: Use a synchronized list:

```
private static ArrayList<Bundle> gCachedExtras = Collections.synchronizedList(new ArrayList<Bundle>());
```

File: src/android/com/adobe/phonegap/push/PushPlugin.java [hunk 2]
```diff
@@ -133,10 +134,13 @@ else if (!savedSenderID.equals(senderID)) {
 
                     }
 
-                    if (gCachedExtras != null) {
+                    if (!gCachedExtras.isEmpty()) {
                         Log.v(LOG_TAG, "sending cached extras");
-                        sendExtras(gCachedExtras);
-                        gCachedExtras = null;
+                        Iterator<Bundle> gCachedExtrasIterator = gCachedExtras.iterator();
```
Comment (macdonst) [inline src/android/com/adobe/phonegap/push/PushPlugin.java]: This is where we could get into a problem if a new push message comes in while iterating over the list. So we need to access the iterator in a synchronized block:

```
synchronized(gCachedExtras) {
    Iterator<Bundle> gCachedExtrasIterator = gCachedExtras.iterator();
    while (gCachedExtrasIterator.hasNext()) {
        sendExtras(gCachedExtrasIterator.next());
    }
}
```

**Comment by mikepsinn** [comment]
@macdonst I added your changes but had to change ArrayList to List because apparently that's what synchronizedList returns.  It seems like it still works, so far.

**Comment by macdonst** [comment]
@mikepsinn thanks for all your hard work on this one.

**Comment by mikepsinn** [comment]
@macdonst Sure!  Copying and pasting is where I'm a Viking! :D

**Comment by macdonst** [comment]
@mikepsinn haha, well you did do the reporting, testing and saw it through so give yourself a pat on the back.

**Comment by mikepsinn** [comment]
Thanks! :D

**Comment by nicoabie** [comment]
Great work @mikepsinn!

**Comment by mikepsinn** [comment]
@nicoabie Thanks for fixing it first!

**Pull request merged**